ssl/ech/ech_ssl_apis.c

```diff
@@ -102,5 +343,38 @@ int SSL_CTX_ech_raw_decrypt(SSL_CTX *ctx,
 
 void SSL_CTX_ech_set_callback(SSL_CTX *ctx, SSL_ech_cb_func f)
 {
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_NULL_PARAMETER);
+        return;
+    }
+    ctx->ext.ech.cb = f;
     return;
 }
+
+int SSL_set1_ech_config_list(SSL *ssl, const uint8_t *ecl, size_t ecl_len)
+{
+    int rv = 0;
+    SSL_CONNECTION *s;
+    OSSL_ECHSTORE *es = NULL;
+    BIO *es_in = NULL;
+
+    s = SSL_CONNECTION_FROM_SSL_ONLY(ssl);
+    if (s == NULL)
+        goto err;
+    if (ecl == NULL || ecl_len == 0) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_PASSED_NULL_PARAMETER);
+        goto err;
+    }
+    if ((es_in = BIO_new_mem_buf(ecl, ecl_len)) == NULL
+        || (es = OSSL_ECHSTORE_new(NULL, NULL)) == NULL
+        || OSSL_ECHSTORE_read_echconfiglist(es, es_in) != 1
+        || SSL_set1_echstore(ssl, es) != 1) {
```

Comment:
could do that, sure, but in looking at it, this seemed better - clients will likely make at most one call to this per TLS connection, so in the end I figured readable code was a better goal

Reply:
OpenSSL has some memory allocator hammering issues related to past technical choices.
As you are adding a large bunch of code, i see no need to add more hammering when it can be avoid easily.

This will only be call per client code, never from server code ?

Reply:
Yes, `SSL_set1_ech_config_list` is only for clients and wouldn't make sense for a server (unless that server is also a client). The input here is a base64 or binary ECHConfigList so parsing those is unlikely to involve much memory (maybe input of ~100 octets, bit more than that of an internal data structure). So I'm not convinced the "hammering" thing applies tbh. Might be best to see what other reviewers conclude in this trade-off between a few lines more code or suffering the added  alloc/free calls of the current code.

Reply:
Would it make sense instead to make the OSSL_ECHSTORE a refcounted object?